zerver/tests/test_decorators.py

```diff
@@ -108,6 +113,92 @@ def test(request, payload=REQ(argument_type="invalid")):
                 pass
             test(request)
 
+    def test_api_key_only_webhook_view(self):
+        @api_key_only_webhook_view('ClientName')
+        def get_user_profile_api_key(request, user_profile, client):
+            return user_profile.api_key
+
+        class Request(object):
+            REQUEST = {} # type: Dict[str, str]
+            COOKIES = {}
+            META = {'PATH_INFO': ''}
+
+        webhook_bot_email = 'webhook-bot@zulip.com'
+        request = Request()
+
+        request.REQUEST['api_key'] = 'not_existing_api_key'
+        with self.assertRaises(JsonableError):
+            get_user_profile_api_key(request)
+
+        request.REQUEST['api_key'] = get_user_profile_by_email(webhook_bot_email).api_key
+        request.COOKIES['csrftoken'] = get_token(request)
+        self.assertEqual(get_user_profile_api_key(request), get_user_profile_by_email(webhook_bot_email).api_key)
+
+    def test_authenticated_json_post_view(self):
+        @authenticated_json_post_view
+        def get_user_profile_api_key(request, user_profile, client):
+            return user_profile.api_key
+
+        webhook_bot_email = 'webhook-bot@zulip.com'
+        user = get_user_profile_by_email(webhook_bot_email)
+
+        class Request(object):
+            REQUEST = {} # type: Dict[str, str]
+            COOKIES = {}
+            META = {'PATH_INFO': ''}
+            method = 'POST'
+
+        request = Request()
+        request.user = user
```

Comment:
Hmm.  If you want to be testing the `json` style auth decorators properly, you should pass the input they expect.  So rather than setting `request.user` (which is an output of the auth process), for the JSON views, you should make your code a subclass of AuthedTestCase and use the `self.login` function to get a valid logged-in cookie (and then you don't need to set `api_key`.